File: src/api/params.rs
//! Shared helpers for reading loosely-typed JSON request bodies the way the
//! original Rails controllers did: absent and `null` are equivalent for most
//! fields, numeric fields also accept numeric strings, and enum fields are
//! matched case-sensitively against their string names.

use chrono::NaiveDateTime;
use serde_json::Value;

use super::error::ApiError;
use super::time;

/// Reads a string field, returning `None` when the key is absent or its value
/// is not a JSON string.
pub fn string_field(body: &Value, key: &str) -> Option<String> {
    body.get(key)
        .and_then(Value::as_str)
        .map(ToString::to_string)
}

/// Whether the key is present at all, even when its value is `null`. Used to
/// tell "not provided" apart from "explicitly set to null".
pub fn touched(body: &Value, key: &str) -> bool {
    body.get(key).is_some()
}

/// Parses an enum field written as a string.
pub fn parse_enum_field(
    body: &Value,
    key: &str,
    parser: fn(&str) -> Option<i32>,
) -> Result<Option<i32>, ApiError> {
    match body.get(key) {
        None | Some(Value::Null) => Ok(None),
        Some(Value::String(value)) => parser(value).map(Some).ok_or_else(ApiError::invalid_value),
        Some(_) => Err(ApiError::invalid_value()),
    }
}

/// Parses an `i32` field that may arrive as a JSON number or a numeric string.
pub fn parse_i32_field(body: &Value, key: &str) -> Result<Option<i32>, ApiError> {
    match body.get(key) {
        None | Some(Value::Null) => Ok(None),
        Some(Value::Number(number)) => number
            .as_i64()
            .and_then(|value| i32::try_from(value).ok())
            .map(Some)
            .ok_or_else(ApiError::invalid_value),
        Some(Value::String(value)) => value
            .trim()
            .parse::<i32>()
            .map(Some)
            .map_err(|_| ApiError::invalid_value()),
        Some(_) => Err(ApiError::invalid_value()),
    }
}

/// Parses a datetime string field. Absent/null stays `None`.
pub fn parse_datetime_field(body: &Value, key: &str) -> Result<Option<NaiveDateTime>, ApiError> {
    match body.get(key) {
        None | Some(Value::Null) => Ok(None),
        Some(Value::String(value)) => time::parse_datetime(value)
            .map(Some)
            .ok_or_else(ApiError::invalid_value),
        Some(_) => Err(ApiError::invalid_value()),
    }
}

File: src/common/params.ts
import {ApiError} from "./errors";
import * as time from "./time";

export type JsonObject = Record<string, unknown>;

/** Reads a string field; `null` when absent or not a JSON string. */
export function stringField(body: JsonObject, key: string): string | null {
    const value = body[key];
    return typeof value === "string" ? value : null;
}

/** Whether the key is present at all, even when its value is `null`. */
export function touched(body: JsonObject, key: string): boolean {
    return Object.prototype.hasOwnProperty.call(body, key);
}

/** Parses an enum field written as a string. */
export function parseEnumField(body: JsonObject, key: string, parser: (value: string) => number | null): number | null {
    const value = body[key];
    if (value === undefined || value === null) {
        return null;
    }
    if (typeof value === "string") {
        const parsed = parser(value);
        if (parsed === null) {
            throw ApiError.invalidValue();
        }
        return parsed;
    }
    throw ApiError.invalidValue();
}

/** Parses an integer field that may arrive as a JSON number or numeric string. */
export function parseI32Field(body: JsonObject, key: string): number | null {
    const value = body[key];
    if (value === undefined || value === null) {
        return null;
    }
    if (typeof value === "number") {
        if (!Number.isInteger(value) || !Number.isSafeInteger(value)) {
            throw ApiError.invalidValue();
        }
        if (value < -2_147_483_648 || value > 2_147_483_647) {
            throw ApiError.invalidValue();
        }
        return value;
    }
    if (typeof value === "string") {
        const parsed = Number.parseInt(value.trim(), 10);
        if (Number.isNaN(parsed) || !/^[+-]?\d+$/.test(value.trim())) {
            throw ApiError.invalidValue();
        }
        return parsed;
    }
    throw ApiError.invalidValue();
}

/** Parses a datetime string field; absent/null stays `null`. */
export function parseDatetimeField(body: JsonObject, key: string): Date | null {
    const value = body[key];
    if (value === undefined || value === null) {
        return null;
    }
    if (typeof value === "string") {
        const parsed = time.parseDatetime(value);
        if (!parsed) {
            throw ApiError.invalidValue();
        }
        return parsed;
    }
    throw ApiError.invalidValue();
}
